From: Lidong Zhong <lzhong@suse.com>
Patch-mainline: Queued in subsystem maintainer repository
Git-repo: https://git.kernel.org/cgit/linux/kernel/git/teigland/linux-dlm.git/
Git-commit: 883854c5457a97190f7b0ee20f03bcd9664fc0c2
References: bnc#881939
Subject: dlm: keep listening connection alive with sctp mode

The connection struct with nodeid 0 is the listening socket,
not a connection to another node.  The sctp resend function
was not checking that the nodeid was valid (non-zero), so it
would mistakenly get and resend on the listening connection
when nodeid was zero.

Signed-off-by: Lidong Zhong <lzhong@suse.com>
Signed-off-by: David Teigland <teigland@redhat.com>
---
 fs/dlm/lowcomms.c |    7 +++++++
 1 file changed, 7 insertions(+)

--- a/fs/dlm/lowcomms.c
+++ b/fs/dlm/lowcomms.c
@@ -619,6 +619,12 @@ static void retry_failed_sctp_send(struc
 	log_print("Retry sending %d bytes to node id %d", len, nodeid);
 
 	con = nodeid2con(nodeid, 0);
+
+	if (!nodeid) {
+		log_print("Shouldn't resend data via listening connection.");
+		return;
+	}
+
 	if (!con) {
 		log_print("Could not look up con for nodeid %d\n",
 			  nodeid);
